Intersect.Client/Interface/Game/Map/MinimapWindow.cs

```diff
@@ -0,0 +1,571 @@
+        {
+            if (player == null)
+            {
+                return;
+            }
+
+            if (!MapInstance.TryGet(player.MapInstance.Id, out var mapBase))
+            {
+                return;
+            }
+
+            var newGrid = CreateMapGridFromMap(mapBase);
+
+            if (!newGrid.SequenceEqual(_mapGrid))
+            {
+                _mapGrid = newGrid;
+                _mapPosition.Clear();
+                foreach (var map in _mapGrid)
+                {
+                    if (map.Value != null)
+                    {
+                        _mapPosition.Add(map.Value.Id, map.Key);
+                    }
+                }
+
+                _redrawMaps = true;
+            }
+
+            var newLocations = GenerateEntityInfo(allEntities, player);
+            if (!newLocations.Equals(_entityInfoCache))
+            {
+                _entityInfoCache = newLocations;
+                _redrawEntities = true;
+            }
+
+            // Update our minimap display area
+            var centerX = (_renderTexture.Width / 3) + (player.X * _minimapTileSize.X);
+            var centerY = (_renderTexture.Height / 3) + (player.Y * _minimapTileSize.Y);
+            var displayWidth = (int)(_renderTexture.Width * (_zoomLevel / 100f));
+            var displayHeight = (int)(_renderTexture.Height * (_zoomLevel / 100f));
+
+            var x = centerX - (displayWidth / 2);
+            if (x + displayWidth > _renderTexture.Width)
+            {
+                x = _renderTexture.Width - displayWidth;
+            }
+
+            if (x < 0)
+            {
+                x = 0;
+            }
+
+            var y = centerY - (displayHeight / 2);
+            if (y + displayHeight > _renderTexture.Height)
+            {
+                y = _renderTexture.Height - displayHeight;
+            }
+
+            if (y < 0)
+            {
+                y = 0;
+            }
+
+            _minimap.SetTextureRect(x, y, displayWidth, displayHeight);
+        }
+
+        private void DrawMinimap()
+        {
+            if (!_redrawEntities && !_redrawMaps)
+            {
+                return;
+            }
+            
+            _renderTexture.Clear(Color.Transparent);
+
+            foreach (var pos in _mapGrid.Keys)
+            {
+                if (_redrawMaps)
+                {
+                    GenerateMinimapCacheFor(pos);
+                }
+
+                if (_redrawEntities)
+                {
+                    GenerateEntityCacheFor(pos);
+                }
+
+                DrawMinimapCacheToTexture(pos);
+            }
+
+            if (_redrawMaps)
+            {
+                _redrawMaps = false;
+            }
+
+            if (_redrawEntities)
+            {
+                _redrawEntities = false;
+            }
+        }
+
+        private void GenerateMinimapCacheFor(MapPosition position)
+        {
+            _minimapCache.TryAdd(position, null);
```

Comment:
Why are you calling `TryAdd()` and then setting it on the next line?

You should be doing

```csharp
if (!_minimapCache.TryGet(position, out var cachedMinimap)) {
    cachedMinimap = GenerateMapRenderTexture();
    cachedMinimap.Clear(Color.Transparent);
    _minimapCache[position] = cachedMinimap;
}
```